test/flush_worker_test.cc

```diff
@@ -0,0 +1,161 @@
+      auto guard = std::make_unique<Envoy::Thread::LockGuard>(lock_);
+      stopped_ = true;
+    }));
+  }
+
+  // Set up expected behaviors when run() is called on dispatcher_.
+  void expectDispatcherRun() {
+    EXPECT_CALL(*dispatcher_, run(_))
+        .WillOnce(Invoke([&](Envoy::Event::DispatcherImpl::RunType type) {
+          // The first dispatcher run is in WorkerImpl::start().
+          ASSERT_EQ(Envoy::Event::DispatcherImpl::RunType::NonBlock, type);
+        }))
+        .WillOnce(Invoke([&](Envoy::Event::DispatcherImpl::RunType type) {
+          // The second dispatcher run is in FlushWorkerImpl::work().
+          ASSERT_EQ(Envoy::Event::DispatcherImpl::RunType::RunUntilExit, type);
+          simulateTimerLoop();
+        }));
+  }
+
+  // Simulate the periodical timer which keeps running until dispatcher->exit()
+  // is called from another thread. The std::promise signal_dispatcher_to_exit_
+  // is used here to guarantee the while loop runs for at least 100 times before dispatcher->exit()
+  // is called.
+  void simulateTimerLoop() {
+    int i = 0;
+    while (true) {
+      i++;
+      // At least run the while loop for 100 times.
+      if (i == 100) {
+        signal_dispatcher_to_exit_.set_value();
+      }
+      auto guard = std::make_unique<Envoy::Thread::LockGuard>(lock_);
+      if (stopped_) {
+        break;
+      } else if (timer_set_) {
+        timer_set_ = false;
+        timer_cb_();
+      }
+    }
+  }
+
+  NiceMock<Envoy::Api::MockApi> mock_api_;
+  Envoy::Thread::ThreadFactory& thread_factory_;
+  Envoy::Stats::IsolatedStoreImpl store_;
+  NiceMock<Envoy::ThreadLocal::MockInstance> tls_;
+  NiceMock<Envoy::Event::MockDispatcher>* dispatcher_ = nullptr; // not owned
+  std::unique_ptr<Envoy::Runtime::ScopedLoaderSingleton> loader_;
+
+  NiceMock<Envoy::Event::MockTimer>* timer_; // not owned
+  Envoy::Event::TimerCb timer_cb_;
+  bool timer_set_{};
+  bool stopped_{false};
+  // Protect stopped_.
+  Envoy::Thread::MutexBasicLockable lock_;
+  std::promise<void> signal_dispatcher_to_exit_;
+
+  Envoy::Stats::MockSink* sink_ = nullptr; // owned by stats_sinks_
+  std::list<std::unique_ptr<Envoy::Stats::Sink>> stats_sinks_;
+};
+
+// Verify stats are flushed periodically until dispatcher->exit() is called from
+// another thread.
+TEST_F(FlushWorkerTest, WorkerFlushStatsPeriodically) {
+  std::chrono::milliseconds stats_flush_interval{10};
+  setupDispatcherTimerEmulation();
+
+  FlushWorkerImpl worker(mock_api_, tls_, store_, stats_sinks_, stats_flush_interval);
+
+  std::thread thread = std::thread([&worker, this] {
+    // Wait for the while loop run at least 100 times in simulateTimerLoop().
+    signal_dispatcher_to_exit_.get_future().wait();
+    worker.exitDispatcher();
+  });
+
+  expectDispatcherRun();
+  // Check flush() is called at least 100 times in simulateTimerLoop().
+  EXPECT_CALL(*sink_, flush(_)).Times(testing::AtLeast(100));
```

Comment:
Can you help me understand the uncertainty around the number of calls? This is similar to what @dubious90 asked - why are we checking for "at least 100"? How is synchronisation working in this test? If we are controlling the timers and the dispatcher - cannot we write it in a way so that we know exactly how many times it gets called?

Tests without strong event based synchronisation tend to get flaky over time and are hard to troubleshoot.

Reply:
Thanks to @oschaaf , updated the test to make the loop run exactly kNumTimerLoops(100) times.

Previously it is not deterministic because the 2 threads are trying to acquire the lock on `stopped_` and the loop will exit only 1 thread get the lock and set  `stopped_ = true`.